openwisp_controller/config/tests/test_vpn.py

```diff
@@ -1041,6 +1049,110 @@ def test_zerotier_config_creation(self, mock_requests):
             self.assertIn('network_id', context_keys)
             self.assertIn('network_name', context_keys)
 
+    @mock.patch(_ZT_GENERATE_IDENTITY_SUBPROCESS)
+    @mock.patch(_ZT_SERVICE_REQUESTS)
+    def test_zerotier_auto_cert_false(self, mock_requests, mock_subprocess):
+        mock_requests.get.side_effect = [
+            # For node status
+            self._get_mock_response(200, response=self._TEST_ZT_NODE_CONFIG)
+        ]
+        mock_requests.post.side_effect = [
+            # For create network
+            self._get_mock_response(200)
+        ]
+        self._set_subprocess_mock(mock_subprocess)
+        device, vpn, template = self._create_zerotier_vpn_template(auto_cert=False)
+        vpnclient_qs = device.config.vpnclient_set
+        self.assertEqual(vpnclient_qs.count(), 1)
+        self.assertEqual(IpAddress.objects.count(), 1)
+        # When auto_cert is 'False', make sure subprocess.run is not called
+        self.assertEqual(mock_subprocess.run.call_count, 0)
+        vpnclient = vpnclient_qs.first()
+        self.assertEqual(vpnclient.member_id, '')
+        self.assertEqual(vpnclient.zt_identity_secret, '')
+        self.assertEqual(vpnclient.private_key, '')
+        self.assertEqual(vpnclient.public_key, '')
+
+    @mock.patch(_ZT_GENERATE_IDENTITY_SUBPROCESS)
+    @mock.patch(_ZT_SERVICE_REQUESTS)
+    def test_ip_deleted_when_vpnclient_deleted(self, mock_requests, mock_subprocess):
+        mock_requests.get.side_effect = [
+            # For node status
+            self._get_mock_response(200, response=self._TEST_ZT_NODE_CONFIG)
+        ]
+        mock_requests.post.side_effect = [
+            # For create network
+            self._get_mock_response(200)
+        ]
+        self._set_subprocess_mock(mock_subprocess)
+        device, vpn, template = self._create_zerotier_vpn_template()
+        self.assertEqual(IpAddress.objects.count(), 2)
+        self.assertEqual(mock_subprocess.run.call_count, 1)
+        vpnclient_qs = device.config.vpnclient_set
+        self.assertEqual(vpnclient_qs.count(), 1)
+        vpnclient_qs.first().delete()
+        self.assertEqual(IpAddress.objects.count(), 1)
+
+    @mock.patch(_ZT_GENERATE_IDENTITY_SUBPROCESS)
+    @mock.patch(_ZT_SERVICE_REQUESTS)
+    def test_ip_deleted_when_device_deleted(self, mock_requests, mock_subprocess):
+        mock_requests.get.side_effect = [
+            # For node status
+            self._get_mock_response(200, response=self._TEST_ZT_NODE_CONFIG)
+        ]
+        mock_requests.post.side_effect = [
+            # For create network
+            self._get_mock_response(200)
+        ]
+        self._set_subprocess_mock(mock_subprocess)
+        device, vpn, template = self._create_zerotier_vpn_template()
+        self.assertEqual(mock_subprocess.run.call_count, 1)
+        self.assertEqual(device.config.vpnclient_set.count(), 1)
+        device.delete()
+        self.assertEqual(IpAddress.objects.count(), 1)
```

Comment:
I don't see `self.assertEqual(IpAddress.objects.count(), 2)` being checked before this